common/service/dynamicconfig/config.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dynamicconfig
+
+import (
+	"errors"
+	"time"
+)
+
+// Client allows fetching values from a dynamic configuration system NOTE: This does not have async
+// options right now. In the interest of keeping it minimal, we can add when requirement arises.
+type Client interface {
+	GetValue(name Key) (interface{}, error)
+	GetValueWithFilters(name Key, filters map[Filter]interface{}) (interface{}, error)
+}
+
+type nopClient struct{}
+
+func (mc *nopClient) GetValue(name Key) (interface{}, error) {
+	return nil, errors.New("unable to find key")
+}
+
+func (mc *nopClient) GetValueWithFilters(
+	name Key, filters map[Filter]interface{},
+) (interface{}, error) {
+	return nil, errors.New("unable to find key")
+}
+
+// NewNopCollection creates a new nop collection
+func NewNopCollection() *Collection {
+	return NewCollection(&nopClient{})
+}
+
+// NewCollection creates a new collection
+func NewCollection(client Client) *Collection {
+	return &Collection{client}
+}
+
+// Collection of values that are in dynamic config
+type Collection struct {
+	client Client
+}
+
+// GetIntPropertyWithTaskList gets property with taskList filter and asserts that it's an integer
+func (c *Collection) GetIntPropertyWithTaskList(key Key, defaultVal int) func(string) int {
+	return func(taskList string) int {
+		filters := make(map[Filter]interface{})
+		filters[TaskListName] = taskList
+		val, err := c.client.GetValueWithFilters(key, filters)
+		if err != nil {
+			return defaultVal
+		}
+		return val.(int)
+	}
+}
+
+// GetIntProperty gets property and asserts that it's an integer
+func (c *Collection) GetIntProperty(key Key, defaultVal int) func() int {
+	return func() int {
+		val, err := c.client.GetValue(key)
+		if err != nil {
+			return defaultVal
+		}
+		return val.(int)
+	}
+}
+
```

Comment:
Each of Get methods repeats the same code snippet. What about:

```
func (c *Collection) GetProperty(key Key, defaultVal interface{}) func() interface{} {
       return func() int {
 		val, err := c.client.GetValue(key)
 		if err != nil {
 			return defaultVal
 		}
 		return val
 	}
}

```
Then Get methods become like:

```
func (c *Collection) GetFloat64Property(key Key, defaultVal float64) func() float64 {
      return func() float64 { 
          return GetProperty(key, defaultVal)().(float64) 
      }
}
```

Reply:
Good idea.

Reply:
With generic functions here and the previous filter one as well, there is an extra cast for converting defaultValue from empty interface to the corresponding type. I wrote some benchmark tests to compare the two and the previous non-generic approach is about 20% faster (260 versus 210 ns). I would favor performance here over reducing duplication since this function could be called quite frequently based on where it's being used. Let me know what you think. Will push my commit with benchmark test so you can take a look.

Reply:
madhurav-C02RQ0RTG8WM:cadence madhuravi$ go test ./common/service/dynamicconfig/ -bench=.
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/common/service/dynamicconfig
BenchmarkGetIntProperty-8        5000000               279 ns/op
BenchmarkGetIntPropertyOld-8    10000000               219 ns/op
PASS
ok      github.com/uber/cadence/common/service/dynamicconfig    4.122s